libs-haskell/da-hs-base/src/DA/Service/Logger/Impl/GCP.hs

```diff
@@ -62,16 +64,20 @@ import Network.HTTP.Simple
 data GCPState = GCPState
     { gcpFallbackLogger :: Lgr.Handle IO
     -- ^ Fallback logger to log exceptions caused by the GCP logging itself.
-    , gcpLogQueue :: Var [LogEntry]
-    -- ^ Unsent logs with the oldest log entry first.
+    , gcpLogChan :: TChan (LogEntry, IO ())
+    -- ^ Unsent logs. The IO action is a finalizer that is run when the log entry
+    -- has been sent successfully.
     , gcpSessionID :: UUID
     -- ^ Identifier for the current session
     , gcpDamlDir :: FilePath
     -- ^ Directory where we store various files such as the amount of
     -- data sent so far.
     , gcpSentDataFileLock :: Lock
-    -- ^ Lock for accessing sendData
-    -- TODO (MK) This doesn’t actually work for concurrent executions.
+    -- ^ Lock for accessing sendData.
+    -- Note that this is not save if there are multiple damlc executables
```

Comment:
save -> safe

Reply:
Writing text without a typechecker is hard.

Reply:
https://github.com/digital-asset/daml/pull/2107